pkg/upgrade/upgrader/jiva_upgrade.go

```diff
@@ -630,6 +672,206 @@ func jivaUpgrade(pvName, openebsNamespace string, utaskObj *utask.UpgradeTask) (
+		var zero int32
+		deployObj, err = deployClient.WithNamespace(j.ns).Get(j.controllerObj.name)
+		if err != nil {
+			return err
+		}
+		deployObj.Namespace = openebsNamespace
+		deployObj.ResourceVersion = ""
+		deployObj.Spec.Replicas = &zero
+		// if target-affinity is set for the pvc them openebs namespace
+		// needs to be added as a bug fix.
+		if deployObj.Spec.Template.Spec.Affinity != nil {
+			if deployObj.Spec.Template.Spec.Affinity.PodAffinity.
+				RequiredDuringSchedulingIgnoredDuringExecution[0].LabelSelector.
+				MatchExpressions[0].Key == "openebs.io/target-affinity" {
+				deployObj.Spec.Template.Spec.Affinity.PodAffinity.
+					RequiredDuringSchedulingIgnoredDuringExecution[0].
+					Namespaces = []string{j.ns}
+			}
+		}
+
+		klog.Infof("creating controller deployment %s in %s namespace", deployObj.Name, openebsNamespace)
+		deployObj, err = deployClient.WithNamespace(openebsNamespace).Create(deployObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the controller service in openebs namespace
+	_, err = serviceClient.WithNamespace(openebsNamespace).
+		Get(j.controllerObj.name+"-svc", metav1.GetOptions{})
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+	// if the service is not found in openebs namespace migrate it
+	if k8serror.IsNotFound(err) {
+		svcList, err := serviceClient.WithNamespace(j.ns).List(metav1.ListOptions{
+			LabelSelector: ctrlSVCLabel,
+		})
+		if err != nil {
+			return err
+		}
+		svcObj := &corev1.Service{}
+		// if service is found in pvc namespace then recreate using it
+		// otherwise create the service using pv
+		if len(svcList.Items) != 0 {
+			svcObj = &svcList.Items[0]
+			svcObj.Namespace = openebsNamespace
+			svcObj.ResourceVersion = ""
+			klog.Infof("removing controller service %s in %s namespace", svcObj.Name, j.ns)
+			err = serviceClient.WithNamespace(j.ns).Delete(svcObj.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			svcObj, err = getTargetSVC(pvName, openebsNamespace)
+			if err != nil {
+				return err
+			}
+			svcObj.Labels["openebs.io/cas-template-name"] = deployObj.Labels["openebs.io/cas-template-name"]
+		}
+		klog.Infof("creating controller service %s in %s namespace", svcObj.Name, openebsNamespace)
+		svcObj, err = serviceClient.WithNamespace(openebsNamespace).Create(svcObj)
+		if err != nil {
+			return err
+		}
+	}
+	err = deployClient.WithNamespace(j.ns).Delete(j.controllerObj.name, &metav1.DeleteOptions{})
+	if err != nil && !k8serror.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+func getTargetSVC(pvName, openebsNamespace string) (*corev1.Service, error) {
+	pvObj, err := pvClient.Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	storageClass := pvObj.Spec.StorageClassName
+	storageClassObj, err := sc.NewKubeClient().Get(storageClass, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	svcObj := &corev1.Service{}
+	svcObj.ObjectMeta = metav1.ObjectMeta{
```

Comment:
how about getting the existing service object and then doing a deep copy, patching the namespace. I guess this logic is written to account for the case where the previous service object is already deleted and then possible job also is restarted and will have to now go ahead and create the new service. 

Can there be an alternate solution like:
- create a temporary service - with a new name and deleting the older cluster ip. 
- delete orig service
- create new service, by copying from temp service and using the target ip (from PV) as the cluster ip
- as part of cleanup, delete the temporary service.